timemachine/cpp/src/rmsd_restraint.cu

```diff
@@ -0,0 +1,222 @@
+    unsigned long long *d_u,  // buffered
+    cudaStream_t stream) {
+
+
+    if(N != N_) {
+        throw std::runtime_error("N_! = N");
+    }
+
+    std::vector<double> h_x(N*3);
+
+    gpuErrchk(cudaMemcpy(&h_x[0], d_x, h_x.size()*sizeof(*d_x), cudaMemcpyDeviceToHost));
+
+    const int B = h_atom_map_.size() / 2;
+
+    Eigen::MatrixXd x1(B,3);
+    Eigen::MatrixXd x2(B,3);
+
+    // compute centroids etc. on the GPU directly
+
+    for(int b=0; b < B; b++) {
+
+        int src_idx = h_atom_map_[b*2+0];
+        x1(b, 0) = h_x[src_idx*3+0];
+        x1(b, 1) = h_x[src_idx*3+1];
+        x1(b, 2) = h_x[src_idx*3+2];
+
+        int dst_idx = h_atom_map_[b*2+1];
+        x2(b, 0) = h_x[dst_idx*3+0];
+        x2(b, 1) = h_x[dst_idx*3+1];
+        x2(b, 2) = h_x[dst_idx*3+2];
+    }
+
+
+    Eigen::Vector3d h_a_centroid = x1.colwise().mean();
+    Eigen::Vector3d h_b_centroid = x2.colwise().mean();
+
+    x1 = x1.rowwise() - h_a_centroid.transpose();
+    x2 = x2.rowwise() - h_b_centroid.transpose();
+
+    Eigen::MatrixXd c = x2.transpose() * x1;
+
+    Eigen::JacobiSVD<Eigen::MatrixXd> svd(c, Eigen::ComputeFullU | Eigen::ComputeFullV);
+    auto s = svd.singularValues();
+
+    double epsilon = 1e-8;
+
+    gpuErrchk(cudaMemset(d_u_buf_, 0, sizeof(*d_u_buf_)));
+    gpuErrchk(cudaMemset(d_du_dx_buf_, 0, N*3*sizeof(*d_du_dx_buf_)));
+
+    if(s[0] < epsilon || s[1] < epsilon || s[2] < epsilon) {
+        return;
+    }
+
+    Eigen::MatrixXd u = svd.matrixU();
+    Eigen::MatrixXd v = svd.matrixV();
+    Eigen::MatrixXd v_t = v.transpose();
+
+    bool is_reflection = u.determinant() * v_t.determinant() < 0.0;
+
+    Eigen::MatrixXd rotation = u * v_t;
+
+    double term = 0;
+    if(is_reflection) {
+        term = (rotation.trace() + 1)/2 - 1;
+    } else {
+        term = (rotation.trace() - 1)/2 - 1;
+    }
+
+    double squared_term = term*term;
+
+    unsigned long long nrg = static_cast<unsigned long long>(llrint(k_*squared_term*FIXED_EXPONENT));
+
+    // backprop'd forces
+    Eigen::MatrixXd eye = Eigen::MatrixXd::Identity(3, 3);
+
+    double squared_term_adjoint = 1.0;
+    double term_adjoint = squared_term_adjoint*2*term;
+    Eigen::MatrixXd r_a = eye*term_adjoint/2;
+    Eigen::MatrixXd u_a = r_a * v_t.transpose();
+    Eigen::MatrixXd v_a_t = u.transpose() * r_a;
+
+    Eigen::MatrixXd smat(3,3);
+    Eigen::MatrixXd smat_inv(3,3);
+    Eigen::MatrixXd F(3,3);
+    for(int i=0; i < 3; i++) {
+        for(int j=0; j < 3; j++) {
+            if(i == j) {
+                F(i, j) = 0;
+                smat(i, j) = s[i];
+                smat_inv(i, j) = 1/s[i];
+                // s_a_mat(i, j) = s_a
+            } else{
+                F(i, j) = 1/(s[j]*s[j] - s[i]*s[i]);
+                smat(i, j) = 0;
+                smat_inv(i, j) = 0;
+            }
+        }
+    }
+
+    Eigen::MatrixXd v_a = v_a_t.transpose();
+
+    Eigen::MatrixXd lhs_1 = u*F.cwiseProduct(u.transpose()*u_a - u_a.transpose()*u)*smat;
```

Comment:
@proteneer You mentioned that this was based on a paper, a comment mentioning of the paper could be useful in the future.

Reply:
Thanks - added the link.